src/core/thread/mle.cpp

```diff
@@ -710,8 +720,17 @@ otError Mle::SetDeviceMode(uint8_t aDeviceMode)
         break;
 
     case OT_DEVICE_ROLE_CHILD:
-        SetStateChild(GetRloc16());
-        SendChildUpdateRequest();
+        if ((mInitialDeviceMode & ModeTlv::kModeFFD) > (aDeviceMode & ModeTlv::kModeFFD) ||
+            (mInitialDeviceMode & ModeTlv::kModeRxOnWhenIdle) > (aDeviceMode & ModeTlv::kModeRxOnWhenIdle))
+        {
+            BecomeDetached();
+        }
+        else
+        {
+            SetStateChild(GetRloc16());
+            SendChildUpdateRequest();
```

Comment:
Since this PR focuses on the child's device mode update after joined the network, would it be better to make it more comprehensive? For example,  when changing from 's' to 'rs', stop the data poll manager for the child, and handle the queued indirect messages for the parent when notified of the mode update via the Child Update Request.